Simplify byte extract: use lowering when expression is a constant

tautschnig opened the pull request:
When none of the earlier simplification rules apply, rewrite the expression to to apply simplification rules for arithmetic and logic expressions.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Simplify byte extract: use lowering when expression is a constant

When none of the earlier simplification rules apply, rewrite the
expression to to apply simplification rules for arithmetic and logic
expressions.

feliperodri commented:
@tautschnig #7331 got merged, so this PR should be ready for a rebase before going through review.

tautschnig committed:
Simplify concatenation: ensure type consistency

We must not end up with single-element concatenation that has a
type different from its operand. Observed on Pointer_byte_extract8 once
enabling further simplification of byte extract.
